### PR TITLE
implement SAYFROM command set (#274) + unify Battle/Channel

### PR DESCRIPTION
**Pull request by silentwings**
#274

### CODE REVIEW

**Comment by silentwings** [comment]
In the long term - yes, I would want to migrate ChanServ commands to server commands. But, I think what I've done is the best course of action for now.

I think ChanServ shouldn't be removed until lobbies have GUIs to handle it, which isn't going happen soon. When migration of ChanServ commands to serve cmds does happen en masse, imo it makes sense to do it at the same time as fixing #58, which will require a tonne of changes to the battle/channel parts of the protocol anyway (partly because battles would then be identified by their channel name, and the integer battle_id would be removed). I'm not going to do that as part of this PR, this can wait for the future. For now I just added some stuff to make sure the existing right-click menu in SL still functions.

ChanServ has a similar syntax to SPADS (which will also need small changes), so its natural from the users point of view to have them make greater use of ChanServ, when making Battles & Channels into the same thing.

I've put a lot of work to make ChanServ into a friendlier system, with sensible error messages and cleaner code. My aim here is that game devs can (as channel/battle founder) use ChanServ to control who they set as autohost operators, whilst moderators have the oversight of who are the channel founders. Then bans/mutes will done server side, and not by SPADS, which is sensible because autohost owners are often long term afk.

**Comment by silentwings** [comment]
Re MUTE* specifically, I deprecated them because the servers "ChanServ" versions should be CHANNELMUTE* (and similarly for others) -- or a move a JSON MUTE cmd that could be integrated with a mod-only global mute; similarly for other cmds. Not sure which I'd favour, but I want to leave this for the future.

**Comment by silentwings** [comment]
CHANGELOG

The chat for each battle now takes place in a channel (-> Battle is derived form Channel)
These channels have names `__battle__1234`, find it with `!battlename <hostname>` to ChanServ.
When a client joins/leaves a battle, they are auto-joined/left to its channel.
 (-> battles can be registered with ChanServ and set a founder, topic, operators, mutes, bans, etc.) 
To speak in a battle, use SAY/SAYEX/SAYFROM into its channel.

-- Changes to protocol -- 
 - added SAYFROM* command set (see above)
 - JOINBATTLE (server) has an extra arg, containing the name of the channel associated to the battle
 - deprecated SAYBATTLE SAYBATTLEEX 
 - removed SAYBATTLEPRIVATE SAYBATTLEPRIVATEEX (unused)
 - deprecated MUTE UNMUTE MUTELIST SETCHANNELKEY 
 - fix #283: now LEFT is sent to a user who leaves a channel, JOINED is sent to a user who joins a channel, deprecated server->client JOIN 

-- Changes to ChanServ --
 - Much nicer command handling / error messages
 - added !battlename
 - added !mute !unmute listmutes 
 - added !forward !unforward !listforwards
 - removed !lock/!unlock, added !setkey instead
 - renamed !spamprotection to !antispam 
---- 

Clients should send the 'u' compatibility flag to signal that they are prepared for the above protocol changes.
Backwards compatibility exists for clients lacking 'u', with a small number of absent new features:
* they will fail to see bridged users as present in battles and channels, although they will get sent special chat messages (in the "old" format) from bridged users and be able to reply to them
* they will fail to see CHANNELTOPIC and CHANNELMESSAGES in battles

**Comment by silentwings** [comment]
The feature set is now complete, but this still needs some testing.

**Comment by TurBoss** [comment]
I'm doing some testing

< BRIDGECLIENTFROM jauriarts MatrixAppService appservice

> SERVERMSG You are now the bridge bot for location 'jauriarts'
appservice@jauriarts
> BRIDGEDCLIENTFROM jauriarts MatrixAppService appservice

< JOINFROM TurBoss matrix test

but i get no response from server

**Comment by abma** [comment]
why are bridged users stored in db?

**Comment by silentwings** [comment]
otherwise they can't be banned (or reliably removed in anyway) from channels

**Comment by silentwings** [comment]
This is ready for merge/review.

File: Client.py [hunk 1]
```diff
@@ -118,6 +127,7 @@ def Handle(self, data):
 			msg_limits = self.floodlimit[self.access]
 		else:
 			msg_limits = self.floodlimit['user']
+		print("< [" + self.username +"] " + data)
```
Comment (gajop) [inline Client.py]: are you sure you want these prints? (one on line 227 as well)

Comment (silentwings) [reply Client.py]: Oh, yes, those debug prints can go now.

Comment (gajop) [reply Client.py]: other than that, maybe squash this into a single commit? it's a lot of commits for the actual changes;
up to you and abma though

**Comment by silentwings** [comment]
I'll deal with the conflicts in a few days time (currently away from home).

**Pull request merged**